File: Strivers_A2Z_DSA_Sheet/3.Learn_Arrays/Easy/Union_of_Two_Sorted_Arrays.cpp
/* #C++
GFG Link: https://www.geeksforgeeks.org/problems/union-of-two-sorted-arrays-1587115621/1?utm_source=youtube&utm_medium=collab_striver_ytdescription&utm_campaign=union-of-two-sorted-arrays.
Problem Statement: Given two sorted arrays of size n and m respectively, find their union. The Union of two arrays can be defined as the common and distinct elements in the two arrays. Return the elements in sorted order.

Example 1:

Input: 
n = 5, arr1[] = {1, 2, 3, 4, 5}  
m = 5, arr2 [] = {1, 2, 3, 6, 7}
Output: 
1 2 3 4 5 6 7
Explanation: 
Distinct elements including both the arrays are: 1 2 3 4 5 6 7.
Example 2:

Input: 
n = 5, arr1[] = {2, 2, 3, 4, 5}  
m = 5, arr2[] = {1, 1, 2, 3, 4}
Output: 
1 2 3 4 5
Explanation: 
Distinct elements including both the arrays are: 1 2 3 4 5.
Example 3:

Input:
n = 5, arr1[] = {1, 1, 1, 1, 1}
m = 5, arr2[] = {2, 2, 2, 2, 2}
Output: 
1 2
Explanation: 
Distinct elements including both the arrays are: 1 2.
Your Task: 
You do not need to read input or print anything. Complete the function findUnion() that takes two arrays arr1[], arr2[], and their size n and m as input parameters and returns a list containing the union of the two arrays.

Expected Time Complexity: O(n+m).
Expected Auxiliary Space: O(n+m).

Constraints:
1 <= n, m <= 105
-109 <= arr1[i], arr2[i] <= 109
*/

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution{
    public:
    //arr1,arr2 : the arrays
    // n, m: size of arrays
    //Function to return a list containing the union of the two arrays. 
    vector<int> findUnion(int arr1[], int arr2[], int n, int m)
    {
        //Your code here
        //return vector with correct order of elements
        vector<int> res;
        int i=0,j=0;
        while( i<n && j<m){
            if(arr1[i]<arr2[j]){
                if(arr1[i]!=arr1[i-1])
                    res.push_back(arr1[i]);
                i++;
            }else if(arr1[i] > arr2[j]){
                if(arr2[j]!=arr2[j-1])
                    res.push_back(arr2[j]);
                j++;
            }else{
                if(arr1[i]!=arr1[i-1] && arr2[j]!=arr2[j-1])
                    res.push_back(arr1[i]);
                i++;j++;
            }
        }
        while(i<n){
            if(arr1[i]!=arr1[i-1])
                res.push_back(arr1[i]);
            i++;
        }
        while(j<m){
            if(arr2[j]!=arr2[j-1])
                res.push_back(arr2[j]);
            j++;
        }
        return res;
    }
};

int main() {
	    int N, M;
	    cin >>N >> M;
	    int arr1[N];
	    int arr2[M];
	    
	    for(int i = 0;i<N;i++){
	        cin >> arr1[i];
	    }
	    
	    for(int i = 0;i<M;i++){
	        cin >> arr2[i];
	    }
	    Solution ob;
	    vector<int> ans = ob.findUnion(arr1,arr2, N, M);
	    for(int i: ans)cout<<i<<' ';
	    cout << endl;	
	return 0;
}
// } Driver Code Ends

File: Strivers_A2Z_DSA_Sheet/3.Learn_Arrays/Easy/Union_of_Two_Sorted_Arrays.py
''' #Python
GFG Link: https://www.geeksforgeeks.org/problems/union-of-two-sorted-arrays-1587115621/1?utm_source=youtube&utm_medium=collab_striver_ytdescription&utm_campaign=union-of-two-sorted-arrays.
Problem Statement: Given two sorted arrays of size n and m respectively, find their union. The Union of two arrays can be defined as the common and distinct elements in the two arrays. Return the elements in sorted order.

Example 1:

Input: 
n = 5, arr1[] = {1, 2, 3, 4, 5}  
m = 5, arr2 [] = {1, 2, 3, 6, 7}
Output: 
1 2 3 4 5 6 7
Explanation: 
Distinct elements including both the arrays are: 1 2 3 4 5 6 7.
Example 2:

Input: 
n = 5, arr1[] = {2, 2, 3, 4, 5}  
m = 5, arr2[] = {1, 1, 2, 3, 4}
Output: 
1 2 3 4 5
Explanation: 
Distinct elements including both the arrays are: 1 2 3 4 5.
Example 3:

Input:
n = 5, arr1[] = {1, 1, 1, 1, 1}
m = 5, arr2[] = {2, 2, 2, 2, 2}
Output: 
1 2
Explanation: 
Distinct elements including both the arrays are: 1 2.
Your Task: 
You do not need to read input or print anything. Complete the function findUnion() that takes two arrays arr1[], arr2[], and their size n and m as input parameters and returns a list containing the union of the two arrays.

Expected Time Complexity: O(n+m).
Expected Auxiliary Space: O(n+m).

Constraints:
1 <= n, m <= 105
-109 <= arr1[i], arr2[i] <= 109
'''

#User function Template for python3

class Solution:
    
    #Function to return a list containing the union of the two arrays.
    def findUnion(self,arr1,arr2,n,m):
        '''
        :param a: given sorted array a
        :param n: size of sorted array a
        :param b: given sorted array b
        :param m: size of sorted array b
        :return:  The union of both arrays as a list
        '''
        # code here 
        
        i=j=0
        res=[]
        while (i<n and j<m):
            if(arr1[i]<arr2[j]):
                if(i==0 or arr1[i]!=arr1[i-1]):
                    res.append(arr1[i])
                i+=1
            elif(arr1[i] > arr2[j]):
                if(j==0 or arr2[j]!=arr2[j-1]):
                    res.append(arr2[j])
                j+=1
            else:
                if(i==0 or (arr1[i] != arr1[i-1] and arr2[j]!=arr2[j-1])):
                    res.append(arr1[i])
                i+=1
                j+=1
        while(i<n):
            if(i==0 or arr1[i]!=arr1[i-1]):
                res.append(arr1[i])
            i+=1
        
        while(j<m):
            if(j==0 or arr2[j]!=arr2[j-1]):
                res.append(arr2[j])
            j+=1
        
        return res


#{ 
 # Driver Code Starts
#Initial Template for Python 3

# Contributed by : Nagendra Jha
# Modified by : Sagar Gupta


if __name__ == '__main__':
        n,m = map(int,input().strip().split())
        a = list(map(int,input().strip().split()))
        b = list(map(int,input().strip().split()))
        ob=Solution()
        li = ob.findUnion(a,b,n,m)
        for val in li:
            print(val, end = ' ')
        print()
# } Driver Code Ends